From 80ff08137eac9c96341cf94a71a6e1f8479c7d9d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Roger=20Pau=20Monn=C3=A9?= <roger.pau@citrix.com>
Date: Wed, 11 Feb 2026 12:20:39 +0100
Subject: xen/kexec: return error code for unknown hypercalls
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Currently do_kexec_op_internal() will return 0 for unknown hypercalls.  Fix
this by returning -EOPNOTSUPP instead.

While there remove the pointless initialization of ret to -EINVAL; it just
makes the code confusing.

Fixes: d046f361dc93 ("Xen Security Modules: XSM")
Reported-by: Andrew Cooper <andrew.cooper3@citrix.com>
Signed-off-by: Roger Pau Monné <roger.pau@citrix.com>
Reviewed-by: Andrew Cooper <andrew.cooper3@citrix.com>
master commit: 1ee8b11c1106dca6b8fe0d54c0e79146bb6545f0
master date: 2026-02-06 20:56:16 +0100

diff --git a/xen/common/kexec.c b/xen/common/kexec.c
index 84fe8c35976e..1a3a20e1c960 100644
--- a/xen/common/kexec.c
+++ b/xen/common/kexec.c
@@ -1217,7 +1217,7 @@ static int do_kexec_op_internal(unsigned long op,
                                 XEN_GUEST_HANDLE_PARAM(void) uarg,
                                 bool compat)
 {
-    int ret = -EINVAL;
+    int ret;
 
     ret = xsm_kexec(XSM_PRIV);
     if ( ret )
@@ -1258,6 +1258,10 @@ static int do_kexec_op_internal(unsigned long op,
     case KEXEC_CMD_kexec_status:
         ret = kexec_status(uarg);
         break;
+
+    default:
+        ret = -EOPNOTSUPP;
+        break;
     }
 
     clear_bit(KEXEC_FLAG_IN_HYPERCALL, &kexec_flags);
